src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNITcpHandle.cs

```diff
@@ -532,44 +531,22 @@ public override void SetAsyncCallbacks(SNIAsyncCallback receiveCallback, SNIAsyn
         /// <returns>SNI error code</returns>
         public override uint SendAsync(SNIPacket packet, SNIAsyncCallback callback = null)
         {
-            SNIPacket newPacket = packet;
-
-            _writeTaskFactory.StartNew(() =>
-            {
-                try
+            Task writeTask = packet.WriteToStreamAsync(_stream);
+            writeTask.ContinueWith((t) =>
                 {
-                    lock (this)
-                    {
-                        packet.WriteToStream(_stream);
-                    }
-                }
-                catch (Exception e)
-                {
-                    SNILoadHandle.SingletonInstance.LastError = new SNIError(SNIProviders.TCP_PROV, SNICommon.InternalExceptionError, e);
-
-                    if (callback != null)
+                    SNIAsyncCallback cb = callback ?? _sendCallback;
+                    uint status = TdsEnums.SNI_SUCCESS;
+                    if (t.IsFaulted)
                     {
-                        callback(packet, TdsEnums.SNI_ERROR);
+                        SNILoadHandle.SingletonInstance.LastError = new SNIError(SNIProviders.TCP_PROV, SNICommon.InternalExceptionError, t.Exception);
+                        status = TdsEnums.SNI_ERROR;
                     }
-                    else
-                    {
-                        _sendCallback(packet, TdsEnums.SNI_ERROR);
-                    }
-
-                    return;
-                }
-
-                if (callback != null)
-                {
-                    callback(packet, TdsEnums.SNI_SUCCESS);
-                }
-                else
-                {
-                    _sendCallback(packet, TdsEnums.SNI_SUCCESS);
+                    cb(packet, status);
                 }
-            });
+            );
```

Comment:
This should pass `, CancellationToken.None, TaskContinuationOptions.DenyChildAttach, TaskScheduler.Default`, and ideally it would also pass ExecuteSynchronously |'d into the continuation options, and ideally pass the necessary state in as the object state rather than closing over it.  But all of that goes away when this is changed to use await.